### PR TITLE
Resize buffer (issue #125)

### PR DESCRIPTION
**Pull request by JoshLambda**
Function to resize the pixel buffer:

pub fn resize_buffer(&mut self, width: u32, height: u32)

The backing texture and swap chain are re-created and the buffer is resized.
The texture format also has to be stored at creation of `Pixels` so that the texture can be re-created with the same format.

I also updated example `imgui-winit` to account for the resized width when `world::draw()`-ing.

This ~~addresses~~ may help with #125.

### CODE REVIEW

**Comment by LoganDark** [comment]
I don't think this is the right solution. `pixels` should not automatically choose a new buffer size for you; you should be the one to specify.

**Comment by JoshLambda** [comment]
> I don't think this is the right solution. `pixels` should not automatically choose a new buffer size for you; you should be the one to specify.

The buffer is copied to the texture no matter what.
It can only be te same size or smaller (rendered with black bands).

If you want a different buffer size, you have to create your own as an off-screen rendering target.
Then it's up to you to copy/scale the part you want to the actual rendering buffer.
This crate could handle that for you, but I'm not sure that's the goal of a simple framebuffer lib.

**Comment by LoganDark** [comment]
> > I don't think this is the right solution. `pixels` should not automatically choose a new buffer size for you; you should be the one to specify.
> 
> The buffer is copied to the texture no matter what.
> It can only be te same size or smaller (rendered with black bands).
> 
> If you want a different buffer size, you have to create your own as an off-screen rendering target.
> Then it's up to you to copy/scale the part you want to the actual rendering buffer.
> This crate could handle that for you, but I'm not sure that's the goal of a simple framebuffer lib.

A simple framebuffer lib should give you the ability to specify the pixel dimensions of the framebuffer. According to your description, the pixel dimensions of the framebuffer are automatically decided when the window resizes. However, `pixels` is meant to have an **explicit** resolution.

File: examples/imgui-winit/Cargo.toml
Comment (LoganDark) [inline examples/imgui-winit/Cargo.toml:20]: Hm, did the later versions of these introduce some sort of regression or unwanted behavior?

**Comment by JoshLambda** [comment]
> A "simple framebuffer lib" as you describe should give you the ability to specify the pixel dimensions of the framebuffer. According to your description, the pixel dimensions of the framebuffer are automatically decided when the window resizes. However, `pixels` is meant to have an **explicit** resolution.
> 
> Please note that the pixel dimensions of the framebuffer are usually _different_ than the physical dimensions of the surface or swap chain, which must be kept in sync with the window, and do not need to be explicitly specified by the programmer.

I agree, but `PixelsBuilder::pixel_aspect_ratio` is currently hardcoded to 1.0 simply because handling a different aspect ratio is not implemented yet. The point of this PR was to be able to resize the content of the window with the window itself (assuming a pixel ratio of 1.0), which is an improvement to the current padding with black bands.

When resizing the window, the buffer size is computed this way:  `capacity = width * height * texture_format_size`
You typically get `width` and `height` dimensions in physical pixels using `window.inner_size()`

So if you want to, you could apply some ratio yourself to `width` and `height` when calling `pixels.resize()`, but I'm not sure it will do what you want it to do.

Maybe if you could give a specific use case I would get a better idea.

**Comment by LoganDark** [comment]
> > A "simple framebuffer lib" as you describe should give you the ability to specify the pixel dimensions of the framebuffer. According to your description, the pixel dimensions of the framebuffer are automatically decided when the window resizes. However, `pixels` is meant to have an **explicit** resolution.
> > Please note that the pixel dimensions of the framebuffer are usually _different_ than the physical dimensions of the surface or swap chain, which must be kept in sync with the window, and do not need to be explicitly specified by the programmer.
> 
> I agree, but `PixelsBuilder::pixel_aspect_ratio` is currently hardcoded to 1.0 simply because handling a different aspect ratio is not implemented yet. The point of this PR was to be able to resize the content of the window with the window itself (assuming a pixel ratio of 1.0), which is an improvement to the current padding with black bands.
> 
> When resizing the window, the buffer size is computed this way: `capacity = width * height * texture_format_size`
> You typically get `width` and `height` dimensions in physical pixels using `window.inner_size()`
> 
> So if you want to, you could apply some ratio yourself to `width` and `height` when calling `pixels.resize()`, but I'm not sure it will do what you want it to do.
> 
> Maybe if you could give a specific use case I would get a better idea.

I'm not talking about aspect ratios either. Pixels in `pixels` are always square, but the framebuffer does not have to have the same size as the window. Yes, the black borders you mention come up when the aspect ratio of the buffer does not match up with the window; this does not need to change. What I think should change is the fact that currently `pixels` decides the new resolution of the framebuffer for you, rather than letting you decide yourself.

What if you want to resize the buffer but the window size has not changed? For example, an emulator with multiple different graphics modes (which is quite common). Right now you have to make `pixels` think the window size has changed, which is not a great solution.

File: examples/imgui-winit/Cargo.toml
Comment (JoshLambda) [reply examples/imgui-winit/Cargo.toml:20]: Yes, the example wouldn't build at all, because winit_input_helper 0.9 requires winit 0.24, while other crates require winit 0.23. So I had to downgrade both of them to avoid a version conflict.

**Comment by JoshLambda** [comment]
> I'm not talking about aspect ratios either. Pixels in `pixels` are always square, but the framebuffer does not have to have the same size as the window. Yes, the black borders you mention come up when the aspect ratio of the buffer does not match up with the window; this does not need to change. What I think should change is the fact that, in your PR, `pixels` will decide the new resolution of the framebuffer for you, rather than letting you decide yourself.
> 
> What if you want to resize the buffer but the window size has not changed? For example, an emulator with multiple different graphics modes (which is quite common). Right now you have to make `pixels` think the window size has changed, which is not a great solution.

Ok, now I understand better. `Pixels` offers only a physical frame buffer. Which means the buffer is the exact representation of the physical pixels from the window. 
As far as I understand - I'm not the maintainer - this crate is meant to be minimalistic, so it's is not a convenience to represent your pixels abstractly and then scale them for you back to the physical space. In this context this is your own job to do so.

I know that it would be useful, as there are lots of operations we can do with buffers. 
But there you have 2 choices, either as I said earlier, creating a different buffer yourself as an intermediate render target and then scale and copy the content to the physical buffer at each update. Or find a more abstract and high level library which would suit your needs.

**Comment by LoganDark** [comment]
> As far as I understand - I'm not the maintainer - this crate is meant to be minimalistic, so it's is not a convenience to represent your pixels abstractly and then scale them for you back to the physical space. In this context this is your own job to do so.

Um... but it is though? That's almost the exact purpose of `pixels`.

Quick ping to @parasyte, because I'm not sure if I can clear this up. As-is, I don't think this change fits with the spirit of `pixels`.

**Comment by parasyte** [comment]
Hi, Thanks for the PR @JoshLambda. I'll have to look over it more closely, probably this weekend.

If we're talking about the _original intention_ of the crate, then that's pretty easy: The pixel buffer was always meant to be fixed sized. The target audience was like NES emulator and Chip-8 authors. These machines only have one screen resolution, so a fixed size screen was perfect.

FWIW, the scaling ratio is intended to support non-square pixel aspect ratios like NES's 8:7 PAR. It's currently unimplemented because I'm kind of lazy. It can be ignored for now, but the short explanation is that it can be used to horizontally (or vertically) stretch the backing pixel buffer while maintaining sharp pixel edges. For this reason, it will never scale both horizontally and vertically at the same time. There are some more technical details and example screenshots on my blog: https://blog.kodewerx.org/2017/09/tiny-mcu-3d-renderer-part-6-camera.html This is unrelated to scaling in general...

Which leads us to the _original essence_ of `pixels`. Which is to provide a crate that abstracts all of the GPU work away from authors who would rather focus on poking some raw pixel values into an array and getting them to appear on screen. Scaling and custom shaders are a nice bonus because the NES screen resolution is way too small to enjoy at 1:1. Custom shaders give users the option to use NTSC filters, high-quality scaling filters, and so on.

I, for one, am happy to extend the capabilities of the crate to include features that can resize the backing pixel buffer at runtime. E.g. that would allow emulators for old computers to change their video mode/screen resolution. And from that perspective, this PR does not appear to align with the feature request in #125.

----

The `resize` method is for changing the presentation texture size to match the surface (called the surface texture in our code and docs). This method could probably be renamed to avoid any confusion with resizing the pixel buffer (and the backing texture). Of course, `resize` takes care of ensuring the presented image always has sharp pixels, even when the surface size is not evenly divisible by the pixel buffer size. (It does this by padding the image with a black border.) The other thing handled by `resize` is (or eventually will be) pixel aspect ratios. Both operations are simple scaling passes, but they need to know the surface size to work correctly.

The new method that is needed to change the backing pixel buffer (e.g. change the emulator video mode) will have to update the `pixels.context.texture` and recreate the scaling renderer for it.

Both textures should be independently resizable, and as long as the presentation texture size matches the surface size, everything should Just Work :tm:. After changing the video mode, a user might drag the window handle to change the screen size to something more comfortable, or even change the window inner size to some multiple of the new video mode. In all cases, the independent textures will ensure the following properties hold:

1. The image will always be presented with sharp pixels at integer scales, possibly with a black border.
2. The pixel aspect ratio will remain fixed, regardless of the presentation texture/window size.
3. The resolution of the pixel buffer (and associated backing texture) can be resized at runtime to fit application needs. (This is the new part that is needed.)

Hope that helps to clear things up in terms of "the exact purpose of `pixels`" and what I imagined a fix for #125 would look like.

**Comment by LoganDark** [comment]
FWIW, the crate I'm working on supports custom aspect ratios and scaling, custom buffer alignments (other than centered), and high-quality scaling (antialiased pixel edges) already. I plan on supporting custom shaders (and post-processing pipelines in general), multiple layers with transparency, and HDR (floating-point) buffers. Fun!

File: examples/imgui-winit/Cargo.toml
Comment (parasyte) [reply examples/imgui-winit/Cargo.toml:20]: @JoshLambda I just merged a patch that addresses the dependency issues in the `imgui-winit` example. Please rebase or merge on the latest HEAD to pick up the patch. It will conflict with this change, so you will want to revert.

Comment (JoshLambda) [reply examples/imgui-winit/Cargo.toml]: @parasyte Done.

**Comment by JoshLambda** [comment]
Ok, so I discovered it only works when PR https://github.com/parasyte/pixels/pull/122 about handling `wgpu::SwapChainError::Outdated` is applied, so I had to include it in this PR.

**Comment by parasyte** [comment]
#122 probably needs to be merged at some point, but right now it copy-pastes code instead of keeping things DRY.

There are a few issues with this PR that need to be resolved though. The first is that it breaks CI because the examples haven't been updated with the breaking API change on the `resize` method.

The second issue is that the `resize` method should not be changed. The state of the PR right now doesn't allow the underlying byte vector and backing texture to be resized independently of the surface and presentation texture (because it overloads the `resize` method). As described earlier, this should be split into two separate methods so that these textures can be resized in isolation of one another. The existing `resize` method should probably be renamed to something else, I don't know, `resize_surface` maybe.

I don't know why recreating the swap chain is necessary for resizing the byte vector and backing texture. You're probably running into something like #121 because you're overloading the window resize. Just a guess.

**Comment by JoshLambda** [comment]
> #122 probably needs to be merged at some point, but right now it copy-pastes code instead of keeping things DRY.

I moved the swap chain creation in it's own function, so it's now DRY. 
It's being used at `Pixels` creation, when resizing the backing texture and when handling outdated swap chain error.
 
> There are a few issues with this PR that need to be resolved though. The first is that it breaks CI because the examples haven't been updated with the breaking API change on the `resize` method.

Yes, I didn't look at other examples yet.
 
> The second issue is that the `resize` method should not be changed. The state of the PR right now doesn't allow the underlying byte vector and backing texture to be resized independently of the surface and presentation texture (because it overloads the `resize` method). As described earlier, this should be split into two separate methods so that these textures can be resized in isolation of one another. The existing `resize` method should probably be renamed to something else, I don't know, `resize_surface` maybe.

Now we have:
```
/// Resize the surface only (behavior unchanged, making this PR backward compatible)

pub fn resize(&mut self, width: u32, height: u32)

/// Resize the buffer and backing texture

pub fn resize_buffer(&mut self, width: u32, height: u32)
```
We can rename them at your convenience.
 
> I don't know why recreating the swap chain is necessary for resizing the byte vector and backing texture. You're probably running into something like #121 because you're overloading the window resize. Just a guess.

Yes, some attempt at rendering seems to happen between the time when the window is resized and the swap chain is re-created. But I don't know how.

File: src/lib.rs
Comment (parasyte) [inline src/lib.rs:201]: Hm, this looks suspicious. It's passing the same size for both the backing texture (first tuple) and surface texture (second tuple). I believe it should be this:

```suggestion
        self.scaling_matrix_inverse = renderers::ScalingMatrix::new(
            (width as f32, height as f32),
            (
                self.surface_size.width as f32,
                self.surface_size.height as f32,
            ),
        )
```

File: src/lib.rs
Comment (parasyte) [inline src/lib.rs:357]: I bet this would look better with `Result::or_else()`...

```suggestion
        let frame = self
            .context
            .swap_chain
            .get_current_frame()
            .or_else(|err| match err {
                wgpu::SwapChainError::Outdated => {
                    // Recreate the swap chain to mitigate race condition on drawing surface resize.
                    // See https://github.com/parasyte/pixels/issues/121
                    builder::create_swap_chain(self);
                    self.context.swap_chain.get_current_frame()
                }
                err => Err(err),
            })
            .map_err(Error::Swapchain)?;
```

File: src/builder.rs
Comment (parasyte) [inline src/builder.rs:272]: Gut feeling is that I would make this a method of the `Pixels` struct instead of a module-level function on `builder`. Thinking I see why you did it this way, though.

There's a call to `create_swap_chain` further up (line 222) that could be removed _iff_ this was a static method that accepted free arguments instead of an instance method... Kind of like it is right now. And that would be fine.

Other options are changing the type of `Context::swap_chain` to `Option<T>` and then unwrapping it everywhere, or trying to create a half-initialized structure using `unsafe` and [`MaybeUninit`](https://doc.rust-lang.org/nightly/std/mem/union.MaybeUninit.html#initializing-a-struct-field-by-field). I'm not a big fan of `unsafe` though. And having to `unwrap()` the swap chain to use it seems unfortunate. Given that, I think I prefer the static method with free args.

**Review by parasyte — COMMENTED** [review]
Here's my first review. It looks like this might end up closing #121 and #122, which is a nice bonus!

File: src/lib.rs
Comment (parasyte) [inline src/lib.rs:205]: The rest of this function looks a lot like code in the builder. This _could_ be DRY'd, too. That might help with maintainability?

File: src/lib.rs
Comment (JoshLambda) [reply src/lib.rs:201]: We are in function `resize()` which is used to resize the surface when the window is resized. So, `width` and `height` are the dimensions of the surface, not the backing texture.
And indeed, 2 lines earlier we have:

        // Update SurfaceTexture dimensions
        self.surface_size.width = width;
        self.surface_size.height = height;

Unless I'm missing something that means both tuples of the matrix function would be the same in your proposal.

Comment (JoshLambda) [reply src/lib.rs]: Good catch, updated.

File: src/lib.rs
Comment (JoshLambda) [reply src/lib.rs]: This looks way better indeed. Updated.

File: src/builder.rs
Comment (JoshLambda) [reply src/builder.rs:272]: I'm missed the useless call at line 222 (the downsize of coding at 4 AM :) ). I'm looking at it.

Comment (JoshLambda) [reply src/builder.rs]: The useless call to `device.create_swap_chain` line 222 has been removed.
Now we have:

- in builder.rs
/// Generic swap chain creation with free arguments
`pub(crate) fn create_swap_chain(device: &mut wgpu::Device, surface: &wgpu::Surface, format: wgpu::TextureFormat,
    width: u32, height: u32, present_mode: wgpu::PresentMode)`

- in lib.rs
/// Method of Pixels, which calls above function with own values
`pub(crate) fn re_create_swap_chain(&mut self) { /* calls create_swap_chain above, with own values */ }`

File: src/lib.rs
Comment (JoshLambda) [reply src/lib.rs:205]: Yes, that makes sense. I'll look at it when the part about swap chain creation is stabilized.

**Comment by JMS55** [comment]
Since it looks like someone else is tackling this, here's my fork that I never finished. Feel free to use whatever you need from it https://github.com/JMS55/pixels-resize-buffer/commit/b9f8008c87666133784199f8dbbe35804a90fc78.

File: src/builder.rs
Comment (parasyte) [inline src/builder.rs:272]: nitpick: This could accept a `SurfaceSize` as a single arg. Both call sites have direct access to one.

File: src/lib.rs
Comment (parasyte) [inline src/lib.rs:387]: nitpick: minor name change to `recreate_swap_chain` would be preferred.

**Review by parasyte — COMMENTED** [review]
The latest changes have addressed most of my concerns. I have a few nitpicks left (I'm not opposed to keeping things how they are; your call). And the last bit to take care of is DRYing the texture and scaling renderer creation in `Pixels::resize_buffer` and `Builder::build`.

File: src/builder.rs
Comment (JoshLambda) [reply src/builder.rs]: Good idea, done.

File: src/lib.rs
Comment (JoshLambda) [reply src/lib.rs]: Of course, done.

**Comment by JoshLambda** [comment]
I made the builder and buffer_resize as DRY as I could. How does it look like now?

**Review by parasyte — APPROVED** [review]
This is awesome. Thanks for putting this together, and also sorry about the delay.

**Comment by JMS55** [comment]
Just to check, nothing needs to be changed with `ScalingRenderer`? When I was working on this, I was under the impression that you had to change some things (the matrix I think?) whenever the backing texture size changes.

**Comment by parasyte** [comment]
The matrix is for resizing the image on the surface texture. The size of the backing texture should be completely independent of the surface texture or how it gets displayed.

**Pull request merged**
